### PR TITLE
UUID and sha256 digest functions

### PR DESCRIPTION
**Pull request by mbodmer**
issue #1018 mentioned a uuid function would be helpful.

Here is my attempt to deliver this. Also I've included sha256 hashing functions for strings and files.

To create a meaningful unittest for uuid() I had to refactor the Test struct in test.rs, I hope this is acceptable. The challenge was to keep the current interface but allow for custom verification. Well, here is my attempt. I'm new to rust, so bear with me.

### CODE REVIEW

**Comment by mbodmer** [comment]
#1011

**Comment by casey** [comment]
Thanks for this PR! There's already `Test::stderr_regex`, and I think it might be cleaner to add a `Test::stdout_regex` to test that the output matches a pattern, instead of adding a new kind of validator. for UUIDs, it should be okay to check that it matches the pattern `/........-....-....-....-............/`

For the hash functions, can you use the implementation from https://github.com/RustCrypto/hashes? I think it's better maintained.

**Comment by mbodmer** [comment]
Hi, thanks for your review. I have now added stdout_regex in the sense of the existing stderr_regex. I agree, much less complexity that way. Also I have replaced the hashing dependency with rust-crypto. Hope this helps.

**Comment by mbodmer** [comment]
I'm not sure whether the general name digest()/digest_file() for the functions is too broad. Any requests for a better more concise name? :-)
eg: sha256()/sha256_file()

**Comment by casey** [comment]
> I'm not sure whether the general name digest()/digest_file() for the functions is too broad. Any requests for a better more concise name? :-) eg: sha256()/sha256_file()

I think I agree, `sha256()` and `sha256_file()` are probably better, so if you're reading a justfile and you don't know the function you can see what it does without having to look up documentation.

**Comment by mbodmer** [comment]
Ok, they are now renamed to the more specific `sha256()`, `sha256_file()`

File: README.md [hunk 1]
```diff
@@ -1016,6 +1016,12 @@ These functions can fail, for example if a path does not have an extension, whic
 
 - `error(message)` - Abort execution and report error `message` to user.
 
+#### UUID and Hash Generation
+
+- `sha256(textdata)` Generates a sha256 hash of the given text data. Example: `sha256("just hello")`
```
Comment (casey) [inline README.md]:
```suggestion
- `sha256(string)` Return the SHA-256 hash of `string` as a hexadecimal string.
```

File: README.md [hunk 2]
```diff
@@ -1016,6 +1016,12 @@ These functions can fail, for example if a path does not have an extension, whic
...
+- `sha256_file(path)` Generates a sha256 hash of the given file. Example: `sha256_file("/path/to/file")`
```
Comment (casey) [inline README.md]:
```suggestion
- `sha256_file(path)` Return the SHA-256 hash of the file at `path` as a hexadecimal string.
```

File: README.md [hunk 3]
```diff
@@ -1016,6 +1016,12 @@ These functions can fail, for example if a path does not have an extension, whic
...
+- `uuid()` - Randomly generates a hyphenated UUID string.
```
Comment (casey) [inline README.md]:
```suggestion
- `uuid()` - Return a randomly generated UUID.
```

File: src/function.rs [hunk 4]
```diff
@@ -247,6 +250,23 @@ fn replace(_context: &FunctionContext, s: &str, from: &str, to: &str) -> Result<
...
+  hasher.update(s);
+  let hash = hasher.finalize();
+  Ok(format!("{:x}", hash))
+}
+
+fn sha256_file(_context: &FunctionContext, path: &str) -> Result<String, String> {
+  use sha2::{Digest, Sha256};
+  let mut hasher = Sha256::new();
+  let mut file = std::fs::File::open(path).unwrap();
+  std::io::copy(&mut file, &mut hasher).unwrap();
```
Comment (casey) [inline src/function.rs]: These unwraps should both be handled. Something like:

```suggestion
  let mut file = std::fs::File::open(path).map_err(|err| format!("Failed to open file at `{}`: {}", path, err))?;
  std::io::copy(&mut file, &mut hasher).map_err(|err| format!("Failed to read file at `{}`: {}", path, err)?;
```

File: tests/functions.rs [hunk 5]
```diff
@@ -476,3 +476,36 @@ fn path_exists_subdir() {
...
+    .justfile("x := sha256('5943ee37-0000-1000-8000-010203040506')")
+    .args(&["--evaluate", "x"])
+    .stdout("2330d7f5eb94a820b54fed59a8eced236f80b633a504289c030b6a65aef58871")
+    .run();
+}
+
+#[test]
+fn sha256_file() {
+  let mut tmpfile = tempfile::NamedTempFile::new().unwrap();
+  writeln!(tmpfile, "just is great").unwrap();
```
Comment (casey) [inline tests/functions.rs]: ❤️

File: tests/functions.rs [hunk 6]
```diff
@@ -476,3 +476,36 @@ fn path_exists_subdir() {
...
+  let tmppath = tmpfile.into_temp_path();
+
+  let justfile_content = format!("x := sha256_file('{}')", tmppath.to_str().unwrap());
+
+  Test::new()
+    .justfile(justfile_content)
```
Comment (casey) [inline tests/functions.rs]: You can use the somewhat odd `tree` function to populate the tempdir with files:

```suggestion
    .justfile("x := sha256_file('file')")
    .tree(tree! {
      file: "just is great",
    })
```

**Review by casey — CHANGES_REQUESTED** [review]
Nice, looking good. I added a few minor comments.

**Comment by mbodmer** [comment]
Thanks for the extensive review, I think I've covered all your issues now, feel free to check again

**Review by casey — CHANGES_REQUESTED** [review]
Nice, looks good, although I just realized that there's a bug. `just` doesn't change the working directory, so you'll need to join the `path` in `sha256_file` with `context.search.working_directory`, i.e. `context.search.working_directory.join(path)`.

Could you also add a test for this? It should fail on the current PR, and then pass once `path` is joined with the working directory. You can use `Test::current_dir` to make just start in a sub-directory of the tmpdir. You can try to do `sha256_file` on a file in tempdir.

**Comment by mbodmer** [comment]
Sorry for force pushing, I've had to rewrite the history to rebase the branch, since you've also modified the lockfile and stupid me checked it in, in a commit together with other changes.
Can you give me more details, I don't understand where, how and what causes the bug you've mentioned.

**Comment by casey** [comment]
Lets say you have the following files:

```
/justfile
/bar
/foo/
```

And you're currently in `/foo/`. `/justfile` contains:

```
baz:
  echo {{sha256_file("bar")}}}
```

You do:

```
$ just
```

The `just` executable looks for the justfile and sees that it's in `/`, so it sets `search.working_directory` to `/`. However, it does not call `env::set_current_dir` with `/`, so the current directory is still `/bar`. So then `sha256_file("bar")` will fail, because there is no `/foo/bar`.

**Comment by mbodmer** [comment]
Ok, I think now I understand, thanks. Will try

File: src/function.rs [hunk 7]
```diff
@@ -247,6 +250,26 @@ fn replace(_context: &FunctionContext, s: &str, from: &str, to: &str) -> Result<
...
+  use sha2::{Digest, Sha256};
+  let mut hasher = Sha256::new();
+  hasher.update(s);
+  let hash = hasher.finalize();
+  Ok(format!("{:x}", hash))
+}
+
+fn sha256_file(_context: &FunctionContext, path: &str) -> Result<String, String> {
+  use sha2::{Digest, Sha256};
+  let justpath = _context.search.working_directory.join(path);
```
Comment (casey) [inline src/function.rs]: The `_` can be removed now that we're using context. The `_` is just to make rust not complain about an unused variable.

```suggestion
fn sha256_file(context: &FunctionContext, path: &str) -> Result<String, String> {
  use sha2::{Digest, Sha256};
  let justpath = context.search.working_directory.join(path);
```

**Review by casey — APPROVED** [review]
Looks good! I went ahead and removed a `_` that is now unnecessary.

**Comment by mbodmer** [comment]
Thanks

**Pull request merged**